sentry/src/routes/campaign.rs

```diff
@@ -17,28 +38,320 @@ pub async fn create_campaign<A: Adapter>(
-    //                 Err(ResponseError::Conflict(
-    //                     "channel already exists".to_string(),
-    //                 ))
-    //             }
-    //             _ => Err(error_response),
-    //         }
-    //     }
-    //     Ok(false) => Err(error_response),
-    //     _ => Ok(()),
-    // }?;
-
-    let create_response = SuccessResponse { success: true };
+    match insert_campaign(&app.pool, &campaign).await {
+        Err(error) => {
+            error!(&app.logger, "{}", &error; "module" => "create_campaign");
+            return Err(ResponseError::Conflict("campaign already exists".to_string()));
+        }
+        Ok(false) => Err(error_response),
+        _ => Ok(()),
+    }?;
 
     Ok(success_response(serde_json::to_string(&campaign)?))
 }
+
+pub async fn update_campaign<A: Adapter>(
+    req: Request<Body>,
+    app: &Application<A>,
+) -> Result<Response<Body>, ResponseError> {
+    let body = hyper::body::to_bytes(req.into_body()).await?;
+
+    let campaign = serde_json::from_slice::<Campaign>(&body)
+        .map_err(|e| ResponseError::FailedValidation(e.to_string()))?;
+
+    let error_response = ResponseError::BadRequest("err occurred; please try again later".to_string());
+
+    // modify Campaign
+
+    match modify_campaign(&app.pool, &campaign, &app.redis).await {
+        Err(error) => {
+            error!(&app.logger, "{:?}", &error; "module" => "update_campaign");
+            return Err(ResponseError::Conflict("Error modifying campaign".to_string()));
+        }
+        Ok(false) => Err(error_response),
+        _ => Ok(()),
+    }?;
+
+    Ok(success_response(serde_json::to_string(&campaign)?))
+}
+
+// TODO: Double check redis calls
+async fn get_spent_for_campaign(redis: &MultiplexedConnection, id: CampaignId) -> Result<UnifiedNum, ResponseError> {
+    let key = format!("spent:{}", id);
+    // campaignSpent tracks the portion of the budget which has already been spent
+    let campaign_spent = match redis::cmd("GET")
+        .arg(&key)
+        .query_async::<_, Option<String>>(&mut redis.clone())
+        .await
+        {
+            Ok(Some(spent)) => {
+                let res = BigNum::from_str(&spent)?;
+                let res = res.to_u64().ok_or_else(|| {
+                    ResponseError::Conflict("Error while converting BigNum to u64".to_string())
+                })?;
+                Ok(UnifiedNum::from_u64(res))
+            },
+            _ => Ok(UnifiedNum::from_u64(0))
+        };
+
+    campaign_spent
+}
+
+async fn get_remaining_for_campaign_from_redis(redis: &MultiplexedConnection, id: CampaignId) -> Result<UnifiedNum, ResponseError> {
+    let key = format!("remaining:{}", id);
+    let remaining = match redis::cmd("GET")
+        .arg(&key)
+        .query_async::<_, Option<String>>(&mut redis.clone())
+        .await
+         {
+            Ok(Some(remaining)) => {
```

Comment:
Leave a todo for this as we are waiting for #412 to get merged.
- [ ] Leave a todo to use `FromStr` (i.e. `.parse()`) to properly parse the `UnifiedNum` from a string redis value. Mention PR #412 & issue #382 and leave an `to_u64().unwrap()`
- [ ] Add to tasks in issue https://github.com/AdExNetwork/adex-validator-stack-rust/issues/382